apps/web/src/pages/brand/tabs/LayoutEditor.tsx

```diff
@@ -57,22 +64,47 @@ export function LayoutEditor({
       if (layout.description) {
         setValue('description', layout?.description);
       }
+      if (layout.variables) {
+        setValue('variables', layout?.variables);
+      }
       if (layout.isDefault != null) {
         setValue('isDefault', layout?.isDefault);
       }
     }
   }, [layout]);
 
-  async function onUpdateLayout(data) {
-    const updatePayload = {
-      name: data.name,
-      content: data.content,
-      description: data.description,
-      isDefault: data.isDefault,
-    };
+  useMemo(() => {
+    const variables = getTemplateVariables(ast.body).filter(
+      ({ name }) => !isReservedVariableName(name)
+    ) as ITemplateVariable[];
+    const arrayFields = [...(variableArray || [])];
+
+    variables.forEach((vari) => {
+      if (!arrayFields.find((field) => field.name === vari.name)) {
+        arrayFields.push(vari);
+      }
+    });
+    arrayFields.forEach((vari, ind) => {
+      if (!variables.find((field) => field.name === vari.name)) {
+        delete arrayFields[ind];
+      }
+    });
```

Comment:
I am struggling to understand this bit. I have a feeling that this could be done with a single loop. Maybe with `arrayFields` being a Set and using `delete`.

Reply:
Yes, it is a replicated code. Currently the hook we have for it is very dependent on a specific form context (the steps template. 
I thought so to a few months ago haha. Worked on it for hours, eventually realized why  it is needed- I'll tell you the problem later when we have more time and I remind myself a bit.